morph_tool/dendrogram.py

```diff
@@ -0,0 +1,176 @@
+            grp.loc[:, 'x'] = jittered_x
+        return grp
+
+    for dendrogram, position in positions.items():
+        post_section = ((synapses[POST_SECTION_ID] == dendrogram.section_id)
+                        & (synapses[TARGET_NODE_ID] == neuron_node_id))
+        synapses.loc[post_section, 'x'] = position[0]
+        synapses.loc[post_section, 'y'] = position[1] + synapses.loc[
+            post_section, POST_SECTION_POS] * dendrogram.height
+        pre_section = ((synapses[PRE_SECTION_ID] == dendrogram.section_id)
+                       & (synapses[SOURCE_NODE_ID] == neuron_node_id))
+        synapses.loc[pre_section, 'x'] = position[0]
+        synapses.loc[pre_section, 'y'] = position[1] + synapses.loc[
+            pre_section, PRE_SECTION_POS] * dendrogram.height
+
+        # jitter too close synapses
+        df = synapses[pre_section | post_section]
+        if len(df) > 1:
+            # pylint: disable=cell-var-from-loop
+            # group by Y coordinate and jitter by X coordinate
+            synapses[pre_section | post_section] = df \
+                .groupby(lambda idx: round(df.loc[idx, 'y'])) \
+                .apply(_jitter_x)
+
+
+def _add_neurite_legend(fig, dendrogram):
+    def neurite_legend(neurite_type):
+        fig.add_trace(graph_objects.Scatter(
+            x=[None], y=[None],
+            name=neurite_type.name,
+            line={'color': TREE_COLOR[neurite_type]}
+        ))
+
+    if dendrogram.neurite_type != NeuriteType.soma:
+        neurite_legend(dendrogram.neurite_type)
+    else:
+        neurite_types = {d.neurite_type for d in [dendrogram] + dendrogram.children}
+        for neurite_type in neurite_types:
+            neurite_legend(neurite_type)
+
+
+def _get_default_neuron_node_id(synapses):
+    node_id_set = synapses[TARGET_NODE_ID].unique()
+    if node_id_set.size > 1:
+        raise ValueError('Please specify `neuron_node_id` parameter.')
+    return node_id_set[0]
+
+
+def _get_synapse_colormap(synapses):
+    color_list = [
+        'rgb(0,100,0)', 'rgb(139,69,19)', 'rgb(112,128,144)', 'rgb(255,215,0)', 'rgb(218,112,214)']
+    node_id_set = synapses[TARGET_NODE_ID].unique()
+    return {id_: color_list[idx % len(color_list)] for idx, id_ in enumerate(node_id_set)}
+
+
+def draw(neuron, synapses=None, neuron_node_id=None):
+    """Draw dendrogram with synapses.
+
+    Args:
+        neuron (Neurite|Neuron): a Neurite or a Neuron instance of NeuroM package.
+        synapses (DataFrame): synapses dataframe.
+        neuron_node_id (int|None): node id of ``neuron``. If None then it is taken from
+            ``synapses[TARGET_NODE_ID]``.
+
+    Returns:
+        plotly.graph_objects.Figure: plotly figure
+    """
+    dendrogram = SynDendrogram(neuron)
+    positions = layout_dendrogram(dendrogram, np.array([0, 0]))
+    w, h = get_size(positions)
+    positions = move_positions(positions, np.array([.5 * w, 0]))
+    if not synapses.empty:
+        if neuron_node_id is None:
```

Comment:
Since `_position_synapses` mutates `synapses` I think you should work on a copy at this point.

Reply:
Thank you! That was a hidden bug. It will be fixed by moving `synapses.astype` inside `draw`